cpp/test/prims/sample_without_replacement.cu

```diff
@@ -0,0 +1,232 @@
+  for (int i = 0; i < params.sampledLen; ++i) {
+    auto val = h_outIdx[i];
+    // indices must be in the given range
+    ASSERT_TRUE(0 <= val && val < params.len)
+      << "out-of-range index @i=" << i << " val=" << val
+      << " sampledLen=" << params.sampledLen;
+    // indices should not repeat
+    ASSERT_TRUE(occurence.find(val) == occurence.end())
+      << "repeated index @i=" << i << " idx=" << val;
+    occurence.insert(val);
+  }
+}
+INSTANTIATE_TEST_CASE_P(SWoRTests, SWoRTestF, ::testing::ValuesIn(inputsf));
+
+typedef SWoRTest<double> SWoRTestD;
+const std::vector<SWoRInputs<double>> inputsd = {
+  {1024, 512, GenPhilox, 1234ULL},
+  {1024, 1024, GenPhilox, 1234ULL},
+  {1024, 512 + 1, GenPhilox, 1234ULL},
+  {1024, 1024 - 1, GenPhilox, 1234ULL},
+  {1024, 512 + 2, GenPhilox, 1234ULL},
+  {1024, 1024 - 2, GenPhilox, 1234ULL},
+  {1024 + 1, 512, GenPhilox, 1234ULL},
+  {1024 + 1, 1024, GenPhilox, 1234ULL},
+  {1024 + 1, 512 + 1, GenPhilox, 1234ULL},
+  {1024 + 1, 1024 + 1, GenPhilox, 1234ULL},
+  {1024 + 1, 512 + 2, GenPhilox, 1234ULL},
+  {1024 + 1, 1024 - 2, GenPhilox, 1234ULL},
+  {1024 + 2, 512, GenPhilox, 1234ULL},
+  {1024 + 2, 1024, GenPhilox, 1234ULL},
+  {1024 + 2, 512 + 1, GenPhilox, 1234ULL},
+  {1024 + 2, 1024 + 1, GenPhilox, 1234ULL},
+  {1024 + 2, 512 + 2, GenPhilox, 1234ULL},
+  {1024 + 2, 1024 + 2, GenPhilox, 1234ULL},
+
+  {1024, 512, GenTaps, 1234ULL},
+  {1024, 1024, GenTaps, 1234ULL},
+  {1024, 512 + 1, GenTaps, 1234ULL},
+  {1024, 1024 - 1, GenTaps, 1234ULL},
+  {1024, 512 + 2, GenTaps, 1234ULL},
+  {1024, 1024 - 2, GenTaps, 1234ULL},
+  {1024 + 1, 512, GenTaps, 1234ULL},
+  {1024 + 1, 1024, GenTaps, 1234ULL},
+  {1024 + 1, 512 + 1, GenTaps, 1234ULL},
+  {1024 + 1, 1024 + 1, GenTaps, 1234ULL},
+  {1024 + 1, 512 + 2, GenTaps, 1234ULL},
+  {1024 + 1, 1024 - 2, GenTaps, 1234ULL},
+  {1024 + 2, 512, GenTaps, 1234ULL},
+  {1024 + 2, 1024, GenTaps, 1234ULL},
+  {1024 + 2, 512 + 1, GenTaps, 1234ULL},
+  {1024 + 2, 1024 + 1, GenTaps, 1234ULL},
+  {1024 + 2, 512 + 2, GenTaps, 1234ULL},
+  {1024 + 2, 1024 + 2, GenTaps, 1234ULL},
+
+  {1024, 512, GenKiss99, 1234ULL},
+  {1024, 1024, GenKiss99, 1234ULL},
+  {1024, 512 + 1, GenKiss99, 1234ULL},
+  {1024, 1024 - 1, GenKiss99, 1234ULL},
+  {1024, 512 + 2, GenKiss99, 1234ULL},
+  {1024, 1024 - 2, GenKiss99, 1234ULL},
+  {1024 + 1, 512, GenKiss99, 1234ULL},
+  {1024 + 1, 1024, GenKiss99, 1234ULL},
+  {1024 + 1, 512 + 1, GenKiss99, 1234ULL},
+  {1024 + 1, 1024 + 1, GenKiss99, 1234ULL},
+  {1024 + 1, 512 + 2, GenKiss99, 1234ULL},
+  {1024 + 1, 1024 - 2, GenKiss99, 1234ULL},
+  {1024 + 2, 512, GenKiss99, 1234ULL},
+  {1024 + 2, 1024, GenKiss99, 1234ULL},
+  {1024 + 2, 512 + 1, GenKiss99, 1234ULL},
+  {1024 + 2, 1024 + 1, GenKiss99, 1234ULL},
+  {1024 + 2, 512 + 2, GenKiss99, 1234ULL},
+  {1024 + 2, 1024 + 2, GenKiss99, 1234ULL}};
+
+TEST_P(SWoRTestD, Result) {
+  std::set<int> occurence;
+  for (int i = 0; i < params.sampledLen; ++i) {
+    auto val = h_outIdx[i];
```

Comment:
To test the weights, we could use an extremely high weight for a few of the items,  spread those high weights spread across the input array, and verify that they ended up at the beginning of the shuffled array (to some epsilon).

Reply:
that's an interesting idea. Let me try that one. Thanks

Reply:
How about now, @cjnolet ?